### PR TITLE
Add support for the global flags and config files

### PR DESCRIPTION
**Pull request by jwhonce**
Note: This PR doesn't provide full rootless support that will be
      addressed in a future PR

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

File: cmd/podmanV2/root.go [hunk 1]
```diff
@@ -83,17 +134,62 @@ func syslogHook() {
...
-		// podman command failed.  Must exit with that exit code
-		// otherwise command exited correctly.
-		registry.SetExitCode(0)
+func rootlessHook() {
+	if rootless.IsRootless() {
+		logrus.Error("rootless mode is currently not supported. Support will return ASAP.")
 	}
-	os.Exit(registry.GetExitCode())
+	// ce, err := registry.NewContainerEngine(rootCmd, []string{})
+	// if err != nil {
```
Comment (baude) [inline cmd/podmanV2/root.go:142]: wanna keep this?

File: pkg/domain/entities/engine.go [hunk 2]
```diff
@@ -19,74 +28,234 @@ func (m EngineMode) String() string {
...
+
+func (c PodmanConfig) DefaultPidsDescription() string {
+	return "Tune container pids limit (set 0 for unlimited)"
+}
+
+func (c PodmanConfig) DefaultDetachKeys() string {
+	return c.Engine.DetachKeys
+}
+
+// // EngineOptions holds the environment for running the engines
```
Comment (baude) [inline pkg/domain/entities/engine.go:157]: want to keep this?

File: pkg/domain/infra/abi/system.go [hunk 3]
```diff
@@ -88,3 +99,146 @@ func (ic *ContainerEngine) VarlinkService(_ context.Context, opts entities.Servi
...
+// users usernamespace.  This function is updated with a  list of commands that
+// should NOT be run within the user namespace.
+func executeCommandInUserNS(cmd *cobra.Command) bool {
+	return os.Geteuid() == 0
+	// if os.Geteuid() == 0 {
+	// 	return false
+	// }
+	// switch cmd {
+	// case _migrateCommand,
+	// 	_mountCommand,
```
Comment (baude) [inline pkg/domain/infra/abi/system.go:185]: want to keep this?

**Comment by baude** [comment]
couple of issues here and there ... no biggies

File: cmd/podmanV2/root.go (hunk 1 above)
Comment (jwhonce) [reply cmd/podmanV2/root.go:142]: will be used in rootless PR

File: pkg/domain/infra/abi/system.go (hunk 3 above)
Comment (jwhonce) [reply pkg/domain/infra/abi/system.go:185]: Will be removed in rootless PR

**Comment by baude** [comment]
LGTM

File: pkg/domain/entities/engine.go (hunk 2 above)
Comment (jwhonce) [reply pkg/domain/entities/engine.go:157]: Will go away with rootless PR

File: cmd/podmanV2/registry/config.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	case "linux":
+		mode = entities.ABIMode
+	default:
+		logrus.Errorf("%s is not a supported OS", runtime.GOOS)
+		os.Exit(1)
+	}
+
+	// TODO: Is there a Cobra way to "peek" at os.Args?
+	for _, v := range os.Args {
+		if strings.HasPrefix(v, "--remote") {
```
Comment (mheon) [inline cmd/podmanV2/registry/config.go:44]: This doesn't seem safe - what if it's `--remote=false`? Why can't we pass the Cobra CLI object in here?

File: cmd/podmanV2/root.go [hunk 5]
```diff
@@ -1,61 +1,112 @@
...
+		TraverseChildren:   true,
+		PersistentPreRunE:  preRunE,
+		RunE:               registry.SubCommandExists,
+		PersistentPostRunE: postRunE,
+		Version:            version.Version,
 	}
 
 	logLevels = entities.NewStringSet("debug", "info", "warn", "error", "fatal", "panic")
-	logLevel  = "error"
+	logLevel  = "warn"
```
Comment (mheon) [inline cmd/podmanV2/root.go:38]: Why? There's a whole lot of warn-level logging we never want to display. I don't see why we'd want to change this convention at this late date.

File: cmd/podmanV2/registry/config.go (hunk 4 above)
Comment (jwhonce) [reply cmd/podmanV2/registry/config.go:44]: This is before the cobra parse has happened.  In the next PR, I'll beef up the check.

File: cmd/podmanV2/root.go (hunk 5 above)
Comment (jwhonce) [reply cmd/podmanV2/root.go:38]: Fixed.

File: cmd/podmanV2/root.go [hunk 6]
```diff
@@ -83,17 +134,62 @@ func syslogHook() {
...
+	flags.StringSliceVar(&cfg.Engine.HooksDir, "hooks-dir", cfg.Engine.HooksDir, "Set the OCI hooks directory path (may be set multiple times)")
+	flags.IntVar(&opts.MaxWorks, "max-workers", 0, "The maximum number of workers for parallel operations")
+	if err := flags.MarkHidden("max-workers"); err != nil {
+		logrus.WithError(err).Warn("unable to mark max-workers flag as hidden")
+	}
+	flags.StringVar(&cfg.Engine.Namespace, "namespace", cfg.Engine.Namespace, "Set the libpod namespace, used to create separate views of the containers and pods on the system")
+	flags.StringVar(&cfg.Engine.StaticDir, "root", "", "Path to the root directory in which data, including images, is stored")
+	flags.StringVar(&opts.Runroot, "runroot", "", "Path to the 'run directory' where all state information is stored")
+	flags.StringSliceVar(&opts.Engine.RuntimePath, "runtime", []string{}, "Path to the OCI-compatible binary used to run containers, default is /usr/bin/runc")
+	// -s is deprecated due to conflict with -s on subcommands
```
Comment (mheon) [inline cmd/podmanV2/root.go:180]: We had a way of handling this in the main codebase, and we're going to have to rediscover it - we need to keep the `-v` short alias for version. But that can happen later.

File: pkg/domain/entities/engine.go [hunk 7]
```diff
@@ -19,74 +28,234 @@ func (m EngineMode) String() string {
...
+	EngineMode  EngineMode
+	Uri         string
+	Identities  []string
+	CGroupUsage string
+	CpuProfile  string
+	MaxWorks    int
+	Syslog      bool
+	Trace       bool
+
+	// Need to map...
```
Comment (mheon) [inline pkg/domain/entities/engine.go:44]: What does this mean?

File: pkg/domain/infra/abi/system.go [hunk 8]
```diff
@@ -88,3 +99,146 @@ func (ic *ContainerEngine) VarlinkService(_ context.Context, opts entities.Servi
 	}
 	return nil
 }
+
+func (ic *ContainerEngine) SetupRootless(cmd *cobra.Command) error {
```
Comment (mheon) [inline pkg/domain/infra/abi/system.go:103]: I'd like an ack from @giuseppe about the rootless configuration functions here, even if they aren't wired up

File: cmd/podmanV2/root.go [hunk 9]
```diff
@@ -83,17 +134,62 @@ func syslogHook() {
...
+	var dummyVersion bool
+	flags.BoolVar(&dummyVersion, "version", false, "Version of Podman")
+
+	cfg := opts.Config
+	flags.StringVar(&cfg.Engine.CgroupManager, "cgroup-manager", cfg.Engine.CgroupManager, opts.CGroupUsage)
+	flags.StringVar(&opts.CpuProfile, "cpu-profile", "", "Path for the cpu profiling results")
+	if err := flags.MarkHidden("cpu-profile"); err != nil {
+		logrus.WithError(err).Warn("unable to mark cpu-profile flag as hidden")
+	}
+	flags.StringSliceVar(&cfg.Engine.ConmonPath, "conmon", []string{}, "Path of the conmon binary")
```
Comment (mheon) [inline cmd/podmanV2/root.go:163]: This should not be a string slice. We accept one conmon path, and only one.

File: cmd/podmanV2/root.go [hunk 10]
```diff
@@ -83,17 +134,62 @@ func syslogHook() {
...
+	flags.StringVar(&cfg.Engine.EventsLogger, "events-backend", cfg.Engine.EventsLogger, `Events backend to use ("file"|"journald"|"none")`)
+	flags.StringSliceVar(&cfg.Engine.HooksDir, "hooks-dir", cfg.Engine.HooksDir, "Set the OCI hooks directory path (may be set multiple times)")
+	flags.IntVar(&opts.MaxWorks, "max-workers", 0, "The maximum number of workers for parallel operations")
+	if err := flags.MarkHidden("max-workers"); err != nil {
+		logrus.WithError(err).Warn("unable to mark max-workers flag as hidden")
+	}
+	flags.StringVar(&cfg.Engine.Namespace, "namespace", cfg.Engine.Namespace, "Set the libpod namespace, used to create separate views of the containers and pods on the system")
+	flags.StringVar(&cfg.Engine.StaticDir, "root", "", "Path to the root directory in which data, including images, is stored")
+	flags.StringVar(&opts.Runroot, "runroot", "", "Path to the 'run directory' where all state information is stored")
+	flags.StringSliceVar(&opts.Engine.RuntimePath, "runtime", []string{}, "Path to the OCI-compatible binary used to run containers, default is /usr/bin/runc")
```
Comment (mheon) [inline cmd/podmanV2/root.go:179]: Should not be a string slice. Only one allowed.

File: cmd/podmanV2/root.go [hunk 11]
```diff
@@ -83,17 +134,62 @@ func syslogHook() {
...
+	flags.BoolVar(&opts.Trace, "trace", false, "Enable opentracing output (default false)")
+	if err := flags.MarkHidden("trace"); err != nil {
+		logrus.WithError(err).Warn("unable to mark trace flag as hidden")
+	}
+
+	// Override default --help information of `--help` global flag
+	var dummyHelp bool
+	flags.BoolVar(&dummyHelp, "help", false, "Help for podman")
+	flags.StringVar(&logLevel, "log-level", logLevel, fmt.Sprintf("Log messages above specified level (%s)", logLevels.String()))
+	flags.BoolVar(&useSyslog, "syslog", false, "Output logging information to syslog as well as the console (default false)")
```
Comment (mheon) [inline cmd/podmanV2/root.go:194]: We should probably disable this for remote lie we do for max-workers and others

Comment (jwhonce) [reply cmd/podmanV2/root.go:194]: I'll add this to rootless PR.

File: cmd/podmanV2/root.go (hunk 10 above)
Comment (jwhonce) [reply cmd/podmanV2/root.go]: fixed

File: cmd/podmanV2/root.go (hunk 9 above)
Comment (jwhonce) [reply cmd/podmanV2/root.go]: fixed

File: pkg/domain/infra/abi/system.go (hunk 8 above)
Comment (jwhonce) [reply pkg/domain/infra/abi/system.go:103]: Next PR will need his review

File: pkg/domain/entities/engine.go (hunk 7 above)
Comment (jwhonce) [reply pkg/domain/entities/engine.go]: Removed.

File: cmd/podmanV2/root.go (hunk 6 above)
Comment (jwhonce) [reply cmd/podmanV2/root.go:172]: I added TODO, so it won't be forgotten

**Pull request merged**